KATA-3270: pod overhead for peer pods reduce actual worker capacity

balintTobik opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
Peer pods' overhead reduce actual worker capacity, however most of the RAM/CPU allocated in cloud instance.

Fixes: [KATA-3270](https://issues.redhat.com/browse/KATA-3270)

**- What I did**

Operator create 'kata-remote' runtime class with a specific and lower overhead, like in upstream.

**- How to verify it**
1. create KataConfig for a peer pods
2. check 'kata-remote' runtime class created a lower overhead than 'kata' runtime class
3. 
**- Description for the changelog**
Use separate pod's overhead related constants for kata and peer pods.

[EDIT from Greg: adapted description to final version]
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

balintTobik committed:
KATA-3270: pod overhead for peer pods reduce actual worker capacity

Signed-off-by: Balint Tobik <btobik@redhat.com>

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -696,14 +694,16 @@ func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName stri
-			Overhead: &nodeapi.Overhead{
+		}
+		// Use same values for Pod Overhead as upstream kata-deploy using, see
+		// https://github.com/kata-containers/packaging/blob/f17450317563b6e4d6b1a71f0559360b37783e19/kata-deploy/k8s-1.18/kata-runtimeClasses.yaml#L7
```
This link points to the old kata-containers upstream repo which is now archived.

Equivalent link in the current repo is :

https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/runtimeclasses/kata-qemu.yaml#L7

Note that this overhead refers to the `kata-qemu` runtime class from upstream (named `kata` runtime class downstream), which is used for running the VMM on the cluster nodes. It accounts for running the kata shim + QEMU + virtiofsd.

In the case of peer pods, the overhead is just the kata shim since the VMM is running in the cloud. Upstream has a different overhead setting for the `kata-remote` runtime class :

https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/runtimeclasses/kata-remote.yaml#L7

These numbers might be questionable but regardless of that, we want to track upstream :
- we must keep keep copies of the upstream values for both kata and kata-remote cases in constants at the top of this file
- the comment above should be put along these constants for reference

balintTobik replied:
Hi @gkurz!
Thanks for clarifying me!

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -696,14 +694,16 @@ func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName stri
+		}
+		// Use same values for Pod Overhead as upstream kata-deploy using, see
+		// https://github.com/kata-containers/packaging/blob/f17450317563b6e4d6b1a71f0559360b37783e19/kata-deploy/k8s-1.18/kata-runtimeClasses.yaml#L7
+		if len(cpuOverhead) > 0 || len(memoryOverhead) > 0 {
```
There will always be an overhead in practice. We will just adjust the numbers. This check is thus not needed.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -2279,8 +2279,8 @@ func (r *KataConfigOpenShiftReconciler) enablePeerPodsMiscConfigs() error {
-	// Create runtimeClass config for peer-pods
-	err = r.createRuntimeClass(peerpodsRuntimeClassName, peerpodsRuntimeClassCpuOverhead, peerpodsRuntimeClassMemOverhead)
+	// Create runtimeClass config for peer-pods without overhead
+	err = r.createRuntimeClass(peerpodsRuntimeClassName, "", "")
```
As explained in the previous comments, we always have an overhead. We just want to use meaningful values for each case.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -79,8 +79,6 @@ const (
 	peerpodsKataRemoteMachineConfigYaml = "mc-40-kata-remote-config.yaml"
 	peerpodsRuntimeClassName            = "kata-remote"
-	peerpodsRuntimeClassCpuOverhead     = "0.25"
-	peerpodsRuntimeClassMemOverhead     = "350Mi"
```
We will actually need to keep these constants but update the values. 

For consistency and improve clarity, we should create similar constants for the regular kata runtime class and use them at https://github.com/openshift/sandboxed-containers-operator/blob/devel/controllers/openshift_controller.go#L1172

See next comment for more details.

gkurz reviewed (commented):
Hi @balintTobik !

Thanks for working on that. I've commented and hopefully clarified what we want to achieved for now, i.e. use different overhead values for `kata` and `kata-remote`.

Also all commits should have a message and a valid Signed-off-by: tag.

Please address the comments in your local repo and force push to this PR.

balintTobik commented:
Hi @gkurz!
Thank you for the review. I've updated the PR according to your comments.

gkurz committed:
Cosmetic fixes

Rever memory overhead of `kata` to 350 megs.

Commit text example :

The pod overhead of peer pods is expected to be lower than
regular kata pods. Use a different set of values for the
`kata` and `kata-remote` runtime classe, like upstream kata
does.

Current numbers for `kata` are kept. New numbers for `kata-remote`
are taken from upstream.The numbers can probably be refined some
more through experiments.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz reviewed (approved):
Hi @balintTobik ! Thanks for this work.

I've pushed another commit on top of your PR that does some trivial polishing and suggests a description for your commit. I will squash at merge time.

Now we'll ask for some feedback from @vvoronko.

littlejawa reviewed (approved):
/lgtm
Thanks!

gkurz commented:
I should have squashed the patches at merge time but I forgot... never mind.
